commons/src/main/java/esa/commons/io/AbstractPathWatcher.java

```diff
@@ -0,0 +1,265 @@
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+abstract class AbstractPathWatcher implements PathWatcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractPathWatcher.class);
+    private final Path path;
+    private final ArrayList<WatchEvent.Kind<?>> events = new ArrayList<>(4);
+    private final Consumer<WatchEventContext<?>> create;
+    private final Consumer<WatchEventContext<?>> delete;
+    private final Consumer<WatchEventContext<?>> modify;
+    private final Consumer<WatchEventContext<?>> overflow;
+    final WatchService watchService;
+    final WatchEvent.Modifier[] modifiers;
+    final WatchEvent.Kind<?>[] kinds;
+
+    private final Executor executor;
+    private final long delay;
+    private final ScheduledExecutorService delayScheduler;
+    private final Set<String> eventSet;
+    private volatile boolean started = false;
+    private volatile boolean stopped = false;
+
+    AbstractPathWatcher(Path path,
+                        Consumer<WatchEventContext<?>> create,
+                        Consumer<WatchEventContext<?>> delete,
+                        Consumer<WatchEventContext<?>> modify,
+                        Consumer<WatchEventContext<?>> overflow,
+                        WatchEvent.Modifier[] modifiers,
+                        long delay,
+                        ScheduledExecutorService delayScheduler) {
+        Checks.checkNotNull(path, "path");
+        this.path = path;
+        this.modifiers = modifiers == null ? new WatchEvent.Modifier[0] : modifiers;
+
+        this.create = doIfConsumerNotNull(create, () -> events.add(StandardWatchEventKinds.ENTRY_CREATE));
+        this.delete = doIfConsumerNotNull(delete, () -> events.add(StandardWatchEventKinds.ENTRY_DELETE));
+        this.modify = doIfConsumerNotNull(modify, () -> events.add(StandardWatchEventKinds.ENTRY_MODIFY));
+        this.overflow = doIfConsumerNotNull(overflow, () -> events.add(StandardWatchEventKinds.OVERFLOW));
+        Checks.checkArg(events.size() > 0, "No processors of WatchEventContext!" +
+                "Please add processors of WatchEventContext by call of on...(),such as onCreate().");
+        this.kinds = events.toArray(new WatchEvent.Kind[0]);
+
+        this.executor = executor(path);
+        this.delay = delay;
+        this.eventSet = delay > 0 ? new ConcurrentHashSet<>() : null;
+        this.delayScheduler = delayScheduler(delayScheduler, path);
+
+        try {
+            this.watchService = FileSystems.getDefault().newWatchService();
+        } catch (IOException e) {
+            throw new WatchException(e);
+        }
+
+        initDir(path);
```

Comment:
init dir after calling `PathWatcher.start()`

Reply:
We expect initDir(path) to occur during the build phase.